Shared feed item page.

caiosba opened the pull request:
## Description

Now clusters (listed in the feed page for a feed that is sharing media) are clickable and take to the feed item page. The feed item page includes a header, a list of workspaces and their claims/fact-checks and a list of media. Both fact-checks and media can be clicked and have details displayed in a modal.

References: CV2-4175, CV2-4176, CV2-4177 and CV2-4178.

## Type of change

- [ ] Performance improvement and/or refactoring (non-breaking change that keeps existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Security mitigation or enhancement
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Automated test (add or update automated tests)

## How has this been tested?

Unit tests added for most new components.

## Checklist

- [x] I have performed a self-review of my own code
- [x] I've made sure my branch is runnable and given good testing steps in the PR description
- [x] I considered secure coding practices when writing this code. Any security concerns are noted above.
- [x] I have commented my code in hard-to-understand areas, if any
- [ ] I have made needed changes to the README
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If I implemented any new components, they are self-contained, their `propTypes` are declared and they use React Hooks and, if data-fetching is required, they use Relay Modern with fragment containers
- [ ] If my components involve user interaction - specifically button, text fields, or other inputs - I have added a [BEM-like class name](https://meedan.atlassian.net/wiki/spaces/ENG/pages/1327628289/Naming+conventions+for+interactive+elements) to the element that is interacted with
- [x] To the best of my knowledge, any new styles are applied according to the design system
- [ ] If I added a new external dependency, I included a rationale for doing so and an estimate of the change in bundle size (e.g., checked in https://bundlephobia.com/)

brianfleming commented on src/app/components/cds/media-cards/Card.js:
```diff
@@ -37,7 +38,7 @@ const Card = ({
     >
       <MaybeLink to={cardUrl}>
-        <div className={styles.cardContent}>
+        <div className={`${styles.cardContent} ${className}`}>
```
FYI @caiosba , super nit here but, you can add a conditional classname this way https://github.com/meedan/check-web/blob/develop/src/app/components/cds/buttons-checkboxes-chips/ButtonMain.js#L27

which will avoid the space if the classname is not provided

caiosba replied:
Will do!

caiosba replied:
Done, thanks for the tip!

brianfleming commented:
Also @caiosba you probably didn't mean to check in the package-lock file

caiosba commented:
> Also @caiosba you probably didn't mean to check in the package-lock file

Yes, good catch, thanks! I just reverted it.

caiosba commented on src/app/components/search/SearchResultsCards/SharedItemCard.js:
Here I'm just extracting the footer (which shows number of media items, number of requests and the date) as a separate component.

caiosba commented on src/app/components/search/SearchResultsCards/SharedItemCardFooter.js:
Not really new, simply extracted from `<SharedItemCard />`.

caiosba commented on src/app/components/feed/FeedClusters.js:
```diff
@@ -209,24 +210,26 @@ const FeedClustersComponent = ({
-                  factCheckCount={cluster.fact_checks_count}
-                  channels={channels.length > 0 && { main: channels[0], others: channels }}
-                />
+                <Link to={`/${team.slug}/feed/${feed.dbid}/item/${cluster.center.dbid}`}>
```
Just making the cluster card clickable, the code inside didn't really change.

caiosba replied:
Good catch! Done!

dariusk commented on src/app/components/feed/FeedClusters.module.css:
```diff
@@ -12,6 +12,14 @@
 
 .feedClusterCard {
   margin-bottom: 16px;
+
```
I think that if you do the above suggestion, we won't need the hover styling here because it'll be already active on `Card`.

caiosba replied:
Confirmed, done!

dariusk commented on src/app/components/feed/FeedItem.js:
```diff
@@ -0,0 +1,116 @@
+    <QueryRenderer
+      environment={Relay.Store}
+      query={graphql`
+        query FeedItemQuery($slug: String!, $feedId: Int!, $projectMediaId: Int!) {
```
I really like seeing a query that's all composable fragments like this!

dariusk commented on src/app/components/feed/FeedItem.js:
```diff
@@ -0,0 +1,116 @@
+  </ErrorBoundary>
+);
+
+FeedItem.proTypes = {
```
Should be `propTypes`

dariusk reviewed (changes requested):
Looks great overall -- particularly I like the `FeedItem`, `FeedItemHeader`, and `FeedItemTeams` components. Really feels like we're using React/graphql the way it's supposed to be used here!

I made a couple change requests, one around a typo and one around implementation of `Link` around `SharedItemCard`

caiosba commented:
Thanks @dariusk , I implemented the changes and re-requested review.

caiosba commented:
Reviewers, this is now ready - please review the changes since your last review.

amoedoamorim commented on src/app/components/annotations/TiplineRequest.js:
```diff
@@ -147,26 +148,30 @@ const TiplineRequest = ({
 };
 
+TiplineRequest.defaultProps = {
+  showButtons: true,
```
I think boolean defaults should always be false. This way we can simply use the prop name when true or omit it as false, instead of passing propName={false}

dariusk replied:
This is a good point. We could change the name of the variable to `hideButtons` and have it be `false` by default. It's a double negative but I think it's more react-idiomatic.

caiosba replied:
Good idea, guys, thanks. I changed it to `hideButtons`.

dariusk commented on src/app/components/media/MediaLanguageSwitcher.js:
```diff
@@ -70,7 +70,7 @@ const MediaLanguageSwitcher = ({ projectMedia, setFlashMessage }) => {
       selectedLanguage={projectMedia.language_code}
       onSubmit={handleUpdate}
-      languages={safelyParseJSON(projectMedia.team?.get_languages)}
+      languages={projectMedia.team ? safelyParseJSON(projectMedia.team.get_languages) : [projectMedia.language_code]}
```
What happens if there's neither `projectMedia.team` nor `projectMedia.language_code`? Do we need to check for that case as well?

caiosba replied:
Hum... sure... I added another fallback... if there's neither `projectMedia.team` nor `projectMedia.language_code`, I'm now falling back to `und`.

dariusk reviewed (changes requested):
Overall looking great. Super declarative code, again with the nice use of fragments, and I like what I saw with `onSeeMore` in the bullet separator too.

I did leave some comments -- mostly style notes or notes on removing some eslint directives, but also wanted to make sure we captured some edge cases that might come up.

dariusk commented on src/app/components/feed/FeedItemMediaDialog.js:
```diff
@@ -0,0 +1,72 @@
+import MediaCardLarge from '../media/MediaCardLarge';
+
+const FeedItemMediaDialog = ({ itemDbid, classes }) => {
+  const { currentTeamSlug, feedDbid, projectMediaDbid } = window.location.pathname.match(/^\/(?<currentTeamSlug>[^/]+)\/feed\/(?<feedDbid>[0-9]+)\/item\/(?<projectMediaDbid>[0-9]+)$/).groups;
```
(Also I prefer having long regular expressions assigned to variable names.)

caiosba replied:
Thanks for pointing this, I thought about this before but really this component should not render in other routes, so, if for any unexpected reason this regexp doesn't match anything, I prefer that it really triggers a Sentry error.

But good point about readability - I ended up change it to:

```jsx
const urlParseRegex = new RegExp('^/(?<currentTeamSlug>[^/]+)/feed/(?<feedDbid>[0-9]+)/item/(?<projectMediaDbid>[0-9]+)$');
const { currentTeamSlug, feedDbid, projectMediaDbid } = urlParseRegex.test(window.location.pathname) && window.location.pathname.match(urlParseRegex).groups;
```

Another advantage of your suggestion to assign the regular expression to a variable using `new RegExp` is that then I don't need to escape the slashes, so it's more readable too.

PS: I prefer to use `regexp.test(string)` when I just want it to return true or false.

amoedoamorim commented on src/app/components/feed/FeedItemTeams.js:
```diff
@@ -0,0 +1,117 @@
+
+        {/* Workspaces part of the cluster (except current one), sorted by number of media items, descending */}
+        {cluster.cluster_teams.edges.map(edge => edge.node).filter(clusterTeam => clusterTeam.team.dbid !== team.dbid).sort((a, b) => (a.media_count < b.media_count) ? 1 : -1).map(clusterTeam => (
+          <div key={clusterTeam.team.dbid}>
```
Unneeded div? Key prop can be used directly in `FeedTeamCard`

caiosba replied:
Good catch! Yes. I just deleted this `<div />`.

amoedoamorim commented on src/app/components/feed/FeedMediaRequests.js:
```diff
@@ -0,0 +1,80 @@
+/* eslint-disable relay/unused-fields */
```
Why are we disabling it here? Adding a comment could guide future refactoring out of it.

caiosba replied:
Thanks, I added a comment with this explanation: We're using the legacy `<Annotations />` component, which doesn't define a fragment.

brianfleming commented on src/app/components/feed/FeedMediaRequests.js:
```diff
@@ -0,0 +1,80 @@
+        noActivityMessage={
+          <FormattedMessage
+            id="feedMediaRequests.noRequest"
+            defaultMessage="0 Requests"
```
fwiw, I went with the `0 Requests` style to match the other things we did on the list view, just for consistency

caiosba replied:
Thanks guys, keeping it.

amoedoamorim reviewed (approved):
Added a few comments, none of them are blockers

caiosba commented on src/app/components/feed/FeedItemMedia.js:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    {/* Media list */}
+    { clusterTeam ? <FeedItemMediaList teamDbid={clusterTeam.team.dbid} /> : null }
```
Ah yes, I like it too... I'm trying to get more used to it. I just pushed this change to 3 places in that code.

caiosba commented on src/app/components/feed/FeedItemMediaList.js:
```diff
@@ -0,0 +1,151 @@
+const FeedItemMediaListComponentWithIntl = injectIntl(FeedItemMediaListComponent);
+
+const FeedItemMediaList = ({ teamDbid }) => {
+  const { currentTeamSlug, feedDbid, projectMediaDbid } = window.location.pathname.match(/^\/(?<currentTeamSlug>[^/]+)\/feed\/(?<feedDbid>[0-9]+)\/item\/(?<projectMediaDbid>[0-9]+)$/).groups;
```
Done!

caiosba commented on src/app/components/feed/FeedTeamCard.js:
```diff
@@ -1,50 +1,156 @@
+  let uncategorizedRequestsCount = clusterTeam.requests_count;
+
+  return (
+    <div // eslint-disable-line jsx-a11y/click-events-have-key-events
```
Thanks for the suggestion and the reference, I applied them, thanks!

caiosba commented on src/app/components/feed/FeedTeamFactCheckDialog.js:
```diff
@@ -0,0 +1,177 @@
+
+const FeedTeamFactCheckDialog = ({ claimDescription, rating, onClose }) => (
+  // Avoid closing the dialog when clicking on it
+  // eslint-disable-next-line jsx-a11y/click-events-have-key-events
```
Thank you, done!

caiosba commented on src/app/components/feed/FeedTeamFactCheckDialog.js:
```diff
@@ -0,0 +1,177 @@
+                        {rating}
+                      </div>
+                    ),
+                    claimDescription.fact_check.language !== 'und' && (
```
Yes, exactly.

caiosba commented:
Thanks @amoedoamorim and @dariusk ! I addressed all the comments and answered all of them.

@dariusk , since you requested changes, can you please re-review?

dariusk reviewed (approved):
Great changes, agree with everything here!
